File: packages/types/src/interaction/tool-approval.ts
/**
 * Tool Approval Types
 * Types related to tool approval requests and responses
 */

/**
 * Pending Tool Call Info
 * Information about a pending tool call for UI display
 */
export interface PendingToolCallInfo {
  /** Tool call ID */
  id: string;
  /** Tool name */
  name: string;
  /** Tool arguments */
  arguments?: string;
  /** Risk level */
  riskLevel?: import("../tool/risk-level.js").ToolRiskLevel;
}

/**
 * Structured Tool Approval Request Data
 * Rich context for tool approval requests
 */
export interface ToolApprovalRequestData {
  /** Tool call ID */
  toolCallId: string;
  /** Tool name */
  toolName: string;
  /** Tool description */
  toolDescription?: string;
  /** Tool parameters */
  parameters: Record<string, unknown>;
  /** Risk level */
  riskLevel?: import("../tool/risk-level.js").ToolRiskLevel;
  /** Pending tools in queue */
  pendingQueue?: PendingToolCallInfo[];
  /** Auto-executed tools results */
  autoExecutedTools?: import("../tool/execution.js").ToolExecutionResult[];
  /** Batch ID */
  batchId?: string;
  /** Tool index in batch */
  toolIndex?: number;
  /** Total tools in batch */
  totalTools?: number;
  
  // Configuration fields (from ToolApprovalOptions)
  /** Approval timeout in milliseconds */
  timeout?: number;
  /** Security preset name */
  securityPreset?: import("../tool/approval.js").SecurityPreset;
}

/**
 * Structured Tool Approval Response Data
 * User's response to tool approval request
 */
export interface ToolApprovalResponseData {
  /** Whether approved */
  approved: boolean;
  /** Edited parameters */
  editedParameters?: Record<string, unknown>;
  /** User instruction */
  userInstruction?: string;
  /** User annotation/comment */
  annotation?: string;
  /** Rejection reason */
  rejectionReason?: string;
  /** Continue with remaining tools in batch */
  continueBatch?: boolean;
}

File: crates/foundation/wf-types/src/interaction/tool_approval.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PendingToolCallInfo {
    pub id: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub arguments: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub risk_level: Option<super::super::tool::ToolRiskLevel>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolApprovalRequestData {
    pub tool_call_id: String,
    pub tool_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_description: Option<String>,
    pub parameters: serde_json::Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub risk_level: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pending_queue: Option<Vec<PendingToolCallInfo>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub batch_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_index: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_tools: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timeout: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub security_preset: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolApprovalResponseData {
    pub approved: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub edited_parameters: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_instruction: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub annotation: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rejection_reason: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub continue_batch: Option<bool>,
}
